### PR TITLE
Project Comments

### PR DESCRIPTION
**Pull request by Nitish145**
#### Describe the changes you have made in this PR -

1. Add threads endpoints.
    - close thread `/api/v1/threads/:id/close`
    - reopen thread `/api/v1/threads/:id/reopen`
    - subscribe thread `/api/v1/threads/:id/subscribe`
    - unsubscribe thread `/api/v1/threads/:id/unsubscribe`

2. Add comments endpoints
    - GET thread's comments `/api/v1/threads/:thread_id/comments`
    - POST new comment `/api/v1/threads/:thread_id/comments`
    - UPDATE comment `/api/v1/comments/:id`
    - Delete, Undelete, Upvote, Downvote, Unvote comment.

3. Format `routes.rb`.
4. Cleanup `base_controller.rb`.
5. Add extensive specs for threads & comments API. 

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by Nitish145** [comment]
@tachyons please review.. onto adding tests..

**Comment by Nitish145** [comment]
@tachyons please have a look, done with adding specs..

**Comment by Nitish145** [comment]
@tachyons

File: app/controllers/api/v1/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal: true
+
+class Api::V1::CommentsController < Api::V1::BaseController
+  before_action :authenticate_user!, except: %i[index]
```
Comment (tachyons) [inline app/controllers/api/v1/comments_controller.rb:4]: What about comment of private projevts/circuits

Comment (Nitish145) [reply app/controllers/api/v1/comments_controller.rb:4]: Can `private_projects` have comments? I preassumed them to not have one.. @tachyons

Comment (tachyons) [reply app/controllers/api/v1/comments_controller.rb:4]: Can you double check ?

Comment (Nitish145) [reply app/controllers/api/v1/comments_controller.rb:4]: Sure @tachyons

Comment (Nitish145) [reply app/controllers/api/v1/comments_controller.rb:4]: No distinction for comments in regard to `project_access_type`, will add required checks soon!!

**Comment by Nitish145** [comment]
Added checks for `:index` action..
1. If `project` is `"Public"`, no need to be authenticated to fetch comments.
2. If `project` is not `"Public"`, user needs to be authenticated as the `author` to fetch comments or `security_transgression`.

please check @tachyons ..

**Comment by Nitish145** [comment]
@tachyons

**Pull request merged**